dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/trackedentityinstance/TrackedEntityInstanceQueryLimitTest.java

```diff
@@ -0,0 +1,172 @@
+
+    private TrackedEntityInstance tei3;
+
+    private TrackedEntityInstance tei4;
+
+    private TrackedEntityType teiType;
+
+    private User user;
+
+    @Override
+    protected void setUpTest()
+        throws Exception
+    {
+        userService = _userService;
+        user = createAndInjectAdminUser();
+
+        orgUnitA = createOrganisationUnit( "A" );
+        organisationUnitService.addOrganisationUnit( orgUnitA );
+
+        user.getOrganisationUnits().add( orgUnitA );
+
+        teiType = createTrackedEntityType( 'P' );
+        trackedEntityTypeService.addTrackedEntityType( teiType );
+
+        program = createProgram( 'P' );
+        programService.addProgram( program );
+
+        tei1 = createTrackedEntityInstance( orgUnitA );
+        tei2 = createTrackedEntityInstance( orgUnitA );
+        tei3 = createTrackedEntityInstance( orgUnitA );
+        tei4 = createTrackedEntityInstance( orgUnitA );
+        tei1.setTrackedEntityType( teiType );
+        tei2.setTrackedEntityType( teiType );
+        tei3.setTrackedEntityType( teiType );
+        tei4.setTrackedEntityType( teiType );
+
+        trackedEntityInstanceService.addTrackedEntityInstance( tei1 );
+        trackedEntityInstanceService.addTrackedEntityInstance( tei2 );
+        trackedEntityInstanceService.addTrackedEntityInstance( tei3 );
+        trackedEntityInstanceService.addTrackedEntityInstance( tei4 );
+
+        pi1 = createProgramInstance( program, tei1, orgUnitA );
+        pi2 = createProgramInstance( program, tei2, orgUnitA );
+        pi3 = createProgramInstance( program, tei3, orgUnitA );
+        pi4 = createProgramInstance( program, tei4, orgUnitA );
+
+        programInstanceService.addProgramInstance( pi1 );
+        programInstanceService.addProgramInstance( pi2 );
+        programInstanceService.addProgramInstance( pi3 );
+        programInstanceService.addProgramInstance( pi4 );
+
+        programInstanceService.enrollTrackedEntityInstance( tei1, program, new Date(), new Date(), orgUnitA );
+        programInstanceService.enrollTrackedEntityInstance( tei2, program, new Date(), new Date(), orgUnitA );
+        programInstanceService.enrollTrackedEntityInstance( tei3, program, new Date(), new Date(), orgUnitA );
+        programInstanceService.enrollTrackedEntityInstance( tei4, program, new Date(), new Date(), orgUnitA );
+
+        userService.addUser( user );
+    }
+
+    @Test
+    void testTeiQueryLimitConfiguredValue()
+    {
+        TrackedEntityInstanceQueryParams params = new TrackedEntityInstanceQueryParams();
+        params.setProgram( program );
+        params.setOrganisationUnits( Sets.newHashSet( orgUnitA ) );
+        params.setOrganisationUnitMode( OrganisationUnitSelectionMode.ALL );
+        params.setUser( user );
+        params.setSkipPaging( true );
+
+        List<Long> teis = trackedEntityInstanceService.getTrackedEntityInstanceIds( params,
+            false, false );
+
+        assertThat( teis, hasSize( is( 3 ) ) );
```

Comment:
Please remove the assertion on the size since the later hamcrest contains assertion will fail if there is an element which is not in your call to `contains`.

For example

```java
        assertThat(Arrays.asList("foo", "bar", "free"), contains("foo", "bar"));
```

leads to

```
java.lang.AssertionError: 
Expected: iterable containing ["foo", "bar"]
     but: not matched: "free"
```

which gives us the necessary context to fix an issue with our test/implementation. If the test were to become flaky/only fail on Github.

With the current assertions the integration test would fail on unequal size of expected/actual collections with no more context.

Reply:
Since we are testing that the limit is 3, would it make sense to assert the size is not >3?  @teleivo 

I would say potentially in this case the contains is more flaky, incase the default ordering is changing (Which we might actually do); Or @zubaira you can specify the order in the test to make sure its deterministic throughout changes like that